v3/src/components/map/utilities/geo-image.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * GeoImage represents a single geographic image and provides methods to process it.
+ */
+export class GeoImage {
+  private img?: HTMLImageElement
+  private canvas?: HTMLCanvasElement
+  private ctx?: CanvasRenderingContext2D
+  private imageData?: Uint8ClampedArray<ArrayBufferLike>
+
+  /**
+   * Loads an image from a URL
+   * @param url - The URL to load the image from
+   * @returns Promise resolving to this GeoImage instance for chaining
+   */
+  public async loadFromUrl(url: string): Promise<GeoImage> {
+    return new Promise((resolve, reject) => {
+      this.img = new Image()
+      this.img.crossOrigin = "anonymous" // Required for canvas operations
+      this.img.onload = () => resolve(this)
+      this.img.onerror = () => reject(new Error(`Failed to load image ${url}`))
+      this.img.src = url
+    })
+  }
+
+  public get width(): number {
+    if (!this.img) {
+      throw new Error("Image not loaded")
+    }
+    // Ensure the image is loaded and has dimensions
+    if (!this.img.complete || !this.img.naturalWidth || !this.img.naturalHeight) {
+      throw new Error("Image not fully loaded or has invalid dimensions")
+    }
+    return this.img.naturalWidth
+  }
+
+  public get height(): number {
+    if (!this.img) {
+      throw new Error("Image not loaded")
+    }
+    // Ensure the image is loaded and has dimensions
+    if (!this.img.complete || !this.img.naturalWidth || !this.img.naturalHeight) {
+      throw new Error("Image not fully loaded or has invalid dimensions")
+    }
+    return this.img.naturalHeight
+  }
+
+  public prepare() {
+    if (!this.img) {
+      throw new Error("Image not loaded")
+    }
+
+    // Ensure the image is loaded and has dimensions
+    if (!this.img.complete || !this.img.naturalWidth || !this.img.naturalHeight) {
+      throw new Error("Image not fully loaded or has invalid dimensions")
+    }
```

Comment:
It makes me 😬 to see so much repeated code. Could these checks be moved into a separate function that gets called in each of these methods?